contracts/USM.sol

```diff
@@ -1,57 +1,48 @@
 pragma solidity ^0.6.7;
 
-import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "./BufferedToken.sol";
 import "@openzeppelin/contracts/math/SafeMath.sol";
-import "./oracles/IOracle.sol";
+import "./WadMath.sol";
 import "@nomiclabs/buidler/console.sol";
+import "./FUM.sol";
 
 /**
  * @title USM Stable Coin
  * @author Alex Roan (@alexroan)
- * @notice Concept by Jacob Eliosoff (@jacob-eliosoff)
+ * @notice Concept by Jacob Eliosoff (@jacob-eliosoff).
  */
-contract USM is ERC20 {
+contract USM is BufferedToken {
     using SafeMath for uint;
+    using WadMath for uint;
 
-    uint constant UNIT = 1e27; // 10**27, MakerDAO's RAY
-    uint constant MINT_FEE = UNIT / 1000; // 0.1%
-    uint constant BURN_FEE = UNIT / 200; // 0.5%
+    uint constant MINT_FEE = WAD / 1000; // 0.1%
+    uint constant BURN_FEE = WAD / 200; // 0.5%
+    uint constant MAX_DEBT_RATIO = 900000000000000000; // 90%
```

Comment:
You can use scientific notation and write it as ` = 9e17`, or stick to WAD and do ` = WAD * 9/10`

Reply:
> From @jacob-eliosoff: I read a bit about wads & rays... So Eth's fixed-point convention is to shift by decimal places, not binary? (Eg in binary MAX_DEBT_RATIO might be 90% * 2^64 = 16602069666338596454, rather than 90% * 10^18 = 900000000000000000.) I'd have thought SHL/SHR make binary a little more appealing but probably just do what's normal... 

Shift operators in solidity work on the binary representation of the number.
Which usually is written in hexadecimal as 0x...
And which is stored in a uint anyway.
uint x = 10; x << 1; returns 20
uint x = 0xa; x << 1; returns 20
So you can define your numbers in decimal, so that you know what you are typing, and then use bit shift operators if you want to speed up sqaure and square root operations

Reply:
FYI, I [updated](https://github.com/jacob-eliosoff/usm-stablecoin/commit/e824c238fdfe6788d92519421bd0f2b694a35a47) usm_constproduct.py's half_exp_approx() to use decimal-shifted inputs rather than binary-shifted - it doesn't affect the algorithm.

Reply:
Cool! Just as a question, would it be very hard to convert this from a recursive to an iterative function?

Reply:
Ha yeah I wavered about that when writing it.  I had done some Googling that suggested recursion in Solidity might not actually be too expensive?  But anyway, as discussed in chat, the right thing here is probably to lean on the `rpow()` you found in ds-math, which basically _is_ the iterative version of the algo I used.  `rpow(x, n)` has the flaw for our purposes that it assumes `n` is an integer, whereas we need to handle fractional exponents, but we can get around that:

```
HALF_TO_THE_TENTH = 933032991536807415981343266

function rhalfexp(uint x) returns (uint z) {                // Approximates 0.5**x
    return rpow(HALF_TO_THE_TENTH, (10 * x + RAY / 2) / RAY)
}
```

Or we could use HALF_TO_THE_HUNDREDTH to get 0.01 input precision in exchange for ~three more iterations inside `rpow()`.  But we don't really need more than 0.1 input precision, the above is probably fine.

Reply:
I've never actually researched recursion in Solidity, so let's give it a go and see what happens!

Reply:
Nah the thing is `rpow()` makes my recursion obsolete!  Recursion might come up again though...